src/Compute/Metadata/GCEMetadataStream.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+/**
+ * Copyright 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+namespace Google\Cloud\Compute\Metadata;
+
+use Google\Cloud\Compute\Metadata\Readers\StreamReader;
+
+/**
+ * A stream wrapper for GCE metadata.
+ *
+ * You can get the GCE metadata values very easily like:
+ *
+ * use Google\Cloud\Compute\Metadata\GCEMetadataStream;
+ *
+ * GCEMetadataStream::register();
+ * $val = file_get_contents('gce-metadata://project/mykey')
+ * or
+ * $val = file_get_contents('gce-metadata://instance/mykey')
+ */
+class GCEMetadataStream
+{
+    /**
+     * The base URL for the metadata.
+     */
+    const METADATA_BASE_URL = 'http://metadata.google.internal/computeMetadata/v1/';
+
+    /**
+     * The current seek position.
+     */
+    private $position;
+
+    /**
+     * The metadata value.
+     */
+    private $val;
+
+    /**
+     * The metadata reader.
+     */
+    private $reader;
+
+    /**
+     * A static method to register this stream wrapper.
+     */
+    public static function register()
+    {
+        stream_wrapper_register(
+            'gce-metadata',
+            '\Google\Cloud\Compute\Metadata\GCEMetadataStream'
+        )
+            or die('Failed to register GCEMetadataStream');
+    }
+
+    public function __construct()
+    {
+        $this->reader = new StreamReader();
+    }
+
+    /**
+     * A method to replace the reader implementation.
+     */
+    public function setReader($reader)
+    {
+        $this->reader = $reader;
+    }
+
+    // @codingStandardsIgnoreStart
+
+    // These methods are mandatory for implementing stream wrappers,
+    // so marking them with ignore tag.
+    //
+    // TODO: Find a way to just suppress certain types of errors with
+    // phpcs.
+    public function stream_open($path, $mode, $options, &$opened_path)
+    {
+        $url = parse_url($path);
+        if ($url['host'] != 'project' and $url['host'] != 'instance') {
+            return false;
+        }
+        $metadata_url = self::METADATA_BASE_URL.$url['host'].'/attributes'
+            .$url['path'];
+        $this->val = $this->reader->read($metadata_url);
+        $this->position = 0;
```

Comment:
Is there a known static list of metadata values? If so, it seems more appropriate to throw an exception when invalid keys requested, i.e:

```php
throw new \Exception('the key you requested is invalid');
```

Since code requesting a non-existent metadata key is an exceptional thing (i.e. no use-case exists for code with invalid metadata keys), this is a more clear/appropriate way to handle it.